kube-runtime/src/events.rs

```diff
@@ -184,74 +224,136 @@ impl Recorder {
-                deprecated_last_timestamp: None,
-                deprecated_source: None,
-                event_time: Some(MicroTime(Utc::now())),
-                regarding: Some(self.reference.clone()),
-                note: ev.note.map(Into::into),
-                metadata: ObjectMeta {
-                    namespace: self.reference.namespace.clone(),
-                    generate_name: Some(format!("{}-", self.reporter.controller)),
-                    ..Default::default()
-                },
-                reporting_controller: Some(self.reporter.controller.clone()),
-                reporting_instance: Some(
-                    self.reporter
-                        .instance
-                        .clone()
-                        .unwrap_or_else(|| self.reporter.controller.clone()),
-                ),
-                series: None,
-                type_: match ev.type_ {
-                    EventType::Normal => Some("Normal".into()),
-                    EventType::Warning => Some("Warning".into()),
-                },
-                related: ev.secondary,
-            })
-            .await?;
+    pub async fn publish(&self, ev: Event, reference: &ObjectReference) -> Result<(), kube_client::Error> {
+        let now = Utc::now();
+
+        let key = self.get_event_key(&ev, reference);
+        let event = match self.events_cache.read().await.get(&key) {
+            Some(e) => {
+                let series = match &e.series {
+                    Some(series) => EventSeries {
+                        count: series.count + 1,
+                        last_observed_time: MicroTime(now),
+                    },
+                    None => EventSeries {
+                        count: 2,
+                        last_observed_time: MicroTime(now),
+                    },
+                };
+                let mut event = e.clone();
+                event.series = Some(series);
+                event
+            }
+            None => self.generate_event(&ev, reference),
+        };
+
+        let events = Api::namespaced(
+            self.client.clone(),
+            reference.namespace.as_ref().unwrap_or(&"default".to_string()),
+        );
+        if event.series.is_some() {
+            events
+                .patch(&event.name_any(), &PatchParams::default(), &Patch::Merge(&event))
+                .await?;
+        } else {
+            events.create(&PostParams::default(), &event).await?;
+        };
+
+        {
+            let mut cache = self.events_cache.write().await;
+            cache.insert(key, event);
+            cache.retain(|_, v| {
+                if let Some(series) = v.series.as_ref() {
+                    series.last_observed_time.0 + EVENT_FINISH_TIME > now
+                } else if let Some(event_time) = v.event_time.as_ref() {
+                    event_time.0 + EVENT_FINISH_TIME > now
+                } else {
+                    true
+                }
```

Comment:
comment suggestion for why there's a retain at the end of a publisher
```suggestion
            // gc past events older than now + CACHE_TTL
            cache.retain(|_, v| {
                if let Some(series) = v.series.as_ref() {
                    series.last_observed_time.0 + CACHE_TTL > now
                } else if let Some(event_time) = v.event_time.as_ref() {
                    event_time.0 + CACHE_TTL > now
                } else {
                    true
                }
```

Reply:
Yep, I put it there because I don't know where else to put it. We have to think of a better place for it.

Reply:
Ok, I realize that in the end doesn't make sense if you have to invalidate a previous event and nothing more was executed. I changed it at the beginning but I would like to run it in a better place and I don't know where.

Also, I added a test in d85f318c38ae83e92b7908473b531aafe340dd71 which I don't like so much but I didn't want to change the API and add a clock to the structure for mocking the time.

Reply:
ah, this is the `event_recorder_cache_retain` test? yeah, that feels a test that is too low (testing internals rather than interface). it would probably be more readable with a call to publish twice without all the internals.

you could perhaps make a mocking time/duration field conditional on `#[cfg(test)]` to avoid mucking with the interface. though it's private anyway.

Reply:
I will take a look when I have time. Thanks for the advice.

Reply:
Okay, I tried it again in 269eed527289080cdce59adf50ac6f47286b955a. If you feel that is good enough, please make a proposal because that's my best currently.

Reply:
It looks good. Appreciate your patience with this issue. Happy to merge after build is green.